SchedulerExtender: add failedPredicateMap in Filter() returns

ping035627 opened the pull request:
Fix #25797. modify extender.Filter for adding extenders information to “failedPredicateMap” in findNodesThatFit.
When all the filtered nodes that passed "predicateFuncs" don’t pass the extenders filter, the failedPredicateMap hasn’t the extenders information, should add it, I think. So when the length of the “filteredNodes.Items” is 0, we can know the integral information. (The length of the “filteredNodes.Items” is 0, may be because the extenders filter failed.)

ping035627 committed:
modify extends.Filter

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

Add docs

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

add docs for ExtenderFilterResult.FailedNodes in types.go

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

Modify the extender.Filter test.

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

Update extender_test.go

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

modify the comments

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

gofmt -s scheduler_interface.go

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

update the comments

Signed-off-by: PingWang <wang.ping5@zte.com.cn>

spxtr commented:
ok to test

hongchaodeng commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -27,7 +27,7 @@ import (
 	// Filter based on extender-implemented predicate functions. The filtered list is
 	// expected to be a subset of the supplied list.
-	Filter(pod *api.Pod, nodes *api.NodeList) (filteredNodes *api.NodeList, err error)
+	Filter(pod *api.Pod, nodes *api.NodeList) (filteredNodes *api.NodeList, failedNodesMap schedulerapi.FailedNodesMap, err error)
```
Docs for 'failedNodesMap'.

hongchaodeng commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,10 +139,14 @@ type ExtenderArgs struct {
 	Nodes api.NodeList `json:"nodes"`
 }
 
+type FailedNodesMap map[string]string
```
Can you add some docs for it?

ping035627 replied:
ok, thanks for reminding.

hongchaodeng commented:
LGTM after adding the docs

hongchaodeng commented:
@davidopp

davidopp commented:
I didn't look carefully at this PR yet but please make sure both of these work
* extender that has been upgraded to use this feature, but master has not been upgraded
* master that has been upgraded, but extender has not

hongchaodeng commented:
> - extender that has been upgraded to use this feature, but master has not been upgraded
> - master that has been upgraded, but extender has not

Right. I am taking back my lgtm.
Here's my suggestion:
- For the first case, it isn't compatible. We need to do some extra work in extender to comply with that.
- For the second case, it's compatible because of json format.

In both cases, it needs test cases. Ping me after these are added and I will review again.

yarggo commented:
"I didn't look carefully at this PR yet but please make sure both of these work

extender that has been upgraded to use this feature, but master has not been upgraded
master that has been upgraded, but extender has not"

Can we add a new interface or API to keep the compliance?

yarggo commented:
I didn't look carefully at this PR yet but please make sure both of these work

extender that has been upgraded to use this feature, but master has not been upgraded
master that has been upgraded, but extender has not

249043822 commented:
Hi ,@hongchaodeng , @davidopp , As for the two cases you said:
.extender that has been upgraded to use this feature, but master has not been upgraded
.master that has been upgraded, but extender has not
I think that both of theses are compatible because of json format, if extender has been upgraded to use this feature and master has not been upgraded, then master will ignore the new field failedNodeMap of responsed body,  the master will process filter as before and vice versa.

hongchaodeng commented:
That's true..
Just in case, I will verify it by testing with real binaries.

hongchaodeng commented:
@ping035627 
Can you change the title to
"SchedulerExtender: add failedPredicateMap in Filter() returns"

hongchaodeng commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -92,12 +92,13 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig, apiVersion string) (al
 
 	if h.filterVerb == "" {
-		return nodes, nil
+		return nodes, schedulerapi.FailedNodesMap{}, nil
```
return `nil`

ping035627 replied:
The failed nodes is relative to the first returned value(filteredNodes), when there are no filterVerb, the filteredNodes is all the nodes, and the failed nodes is a empty map, not nil, I think.
And there are the same way in findNodesThatFit function. I think it's the same consideration.

hongchaodeng commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,10 +139,15 @@ type ExtenderArgs struct {
 }
 
+// FailedNodesMap represents the filtered out nodes map with node name and failed message
+type FailedNodesMap map[string]string
```
Please don't put this in api/types.go

My suggestion is to just use raw type `map[string]string`

hongchaodeng replied:
Sorry. I think it's fine here because it's scheduler/api/.

hongchaodeng commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,10 +139,15 @@ type ExtenderArgs struct {
 type ExtenderFilterResult struct {
 	// Filtered set of nodes where the pod can be scheduled
 	Nodes api.NodeList `json:"nodes,omitempty"`
+	// Filtered out nodes where the pod can't be scheduled
```
Add "and the extender that failed it"

hongchaodeng commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -167,10 +167,13 @@ func findNodesThatFit(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.No
 				return api.NodeList{}, FailedPredicateMap{}, err
 			}
+			for failedNodeName, failedMsg := range failedMap {
+				failedPredicateMap[failedNodeName] = failedMsg
```
What if there are multiple extenders overriding the same node name?

ping035627 replied:
It's no problem because  the "nodes" parameter of Filter is  the filtered  nodes  of  the  last extender, the  next extender  will  filter  from  it.

ravigadde replied:
Ideally failedPredicateMap should be failedNodeMap as a node can fail because of a predicate or an extender.

ping035627 commented:
@hongchaodeng 
I have changed the title and the docs. But the Jenkins unit/integration failed, the error information isn't my changes, can you help me to see the reason? Thanks very much！

W0629 19:37:59.890979   13163 iptables.go:144] Error checking iptables version, assuming version at least 1.4.11: executable file not found in $PATH
W0629 19:37:59.891318   13163 iptables.go:144] Error checking iptables version, assuming version at least 1.4.11: executable file not found in $PATH

W0629 19:38:00.240194   13163 server.go:697] Failed to update /proc/sys/kernel/keys/root_maxkeys: open /proc/sys/kernel/keys/root_maxkeys: read-only file system
I0629 19:38:00.240245   13163 server.go:710] Setting keys bytes in /proc/sys/kernel/keys/root_maxbytes to 25000000
W0629 19:38:00.240269   13163 server.go:713] Failed to update /proc/sys/kernel/keys/root_maxbytes: open /proc/sys/kernel/keys/root_maxbytes: read-only file system
I0629 19:38:00.240291   13163 server.go:726] Started kubelet v1.4.0-alpha.0.1082+2c37fcc215c2c5
E0629 19:38:00.240360   13163 kubelet.go:930] Image garbage collection failed: unable to find data for container /

......

hongchaodeng commented:
@ping035627 
Can you try rebase and push it again? The failure doesn't look related to scheduler.

hongchaodeng commented:
I have verified that even if extender writes the new struct `ExtenderFilterResult`, scheduler can still pass despite not reading the new field.
@ping035627 Can you guys try to test it in your live cluster?
Other than that, I'm fine with the change. Delay to @davidopp

ping035627 commented:
I have rebase and push it again, but the Jenkins unit/integration still fail.

ping035627 commented:
@hongchaodeng , @davidopp , I have modified and passed the test, can give it the LGTM label? Thanks.

ping035627 commented:
@hongchaodeng , @davidopp , I have modified and passed the test, can give it the LGTM label? Thanks.

ping035627 commented:
@davidopp
can you help review this PR? thanks.

ping035627 commented:
ok, I have rebased.

ping035627 commented:
@davidopp
can you help review this PR? thanks.

ping035627 commented:
@davidopp
can you help review this PR? thanks.

davidopp commented:
@ravigadde can you review this?

ravigadde commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -26,8 +26,9 @@ import (
-	// expected to be a subset of the supplied list.
-	Filter(pod *api.Pod, nodes *api.NodeList) (filteredNodes *api.NodeList, err error)
+	// expected to be a subset of the supplied list. The returned failedNodesMap are added to
+	// list the filtered out nodes information.
```
Would like to modify the comment to:

"failedNodesMap optionally contains the list of failed nodes and failure reasons."

This covers the upgrade cases mentioned earlier.

ravigadde commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,10 +139,15 @@ type ExtenderArgs struct {
 	Nodes api.NodeList `json:"nodes"`
 }
 
+// FailedNodesMap represents the filtered out nodes map with node name and failed message
```
represents the filtered out nodes, with node name and failure message

ravigadde commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,10 +139,15 @@ type ExtenderArgs struct {
 type ExtenderFilterResult struct {
 	// Filtered set of nodes where the pod can be scheduled
 	Nodes api.NodeList `json:"nodes,omitempty"`
+	// Filtered out nodes where the pod can't be scheduled and the extender that failed it
```
Nodes where the pod can't be scheduled and the failure message

ravigadde commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -139,10 +139,15 @@ type ExtenderArgs struct {
 }
 
+// FailedNodesMap represents the filtered out nodes map with node name and failed message
+type FailedNodesMap map[string]string
```
Same as above

ravigadde commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -139,10 +139,15 @@ type ExtenderArgs struct {
 type ExtenderFilterResult struct {
 	// Filtered set of nodes where the pod can be scheduled
 	Nodes apiv1.NodeList `json:"nodes,omitempty"`
+	// Filtered out nodes where the pod can't be scheduled and the extender that failed it
```
Same as above

ravigadde commented:
@ping035627 @davidopp Sorry for the delay in reviewing. Added some minor comments. Changes look good to me otherwise.

ping035627 commented:
@ravigadde , @davidopp 
Thanks for review, I've changed the comments referring to your suggestion and passed the test, and now can you continue the process? need LGTM, thanks.

davidopp commented:
@ravigadde can you verify that your comments were addressed, and then I will add LGTM?

ping035627 commented:
@ravigadde can you help to verify that your comments were addressed? thanks!

ping035627 commented:
@ravigadde , @davidopp can you help to verify that the comments were addressed and continue the process? thanks very much.

1) for scheduler_interface.go
"failedNodesMap optionally contains the list of failed nodes and failure reasons."
// expected to be a subset of the supplied list. failedNodesMap optionally contains
// the list of failed nodes and failure reasons.

2) the next two for api/types.go and api/v1/types.go
"represents the filtered out nodes, with node name and failure message"
// FailedNodesMap represents the filtered out nodes, with node name and failure message

"Nodes where the pod can't be scheduled and the failure message"
// Filtered out nodes where the pod can't be scheduled and the failure message

davidopp commented:
I haven't forgotten this PR but I'm extremely backed up at the moment, very sorry.

ping035627 commented:
@apelisse  this pr make "do-not-merge"  by  robot, mis-label?

ravigadde commented:
@ping035627 @davidopp Sorry I was out for a while. I am reviewing it again now.

ravigadde commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,14 +139,14 @@ type ExtenderArgs struct {
 	// Filtered set of nodes where the pod can be scheduled
 	Nodes api.NodeList `json:"nodes,omitempty"`
-	// Filtered out nodes where the pod can't be scheduled and the extender that failed it
+	// Filtered out nodes where the pod can't be scheduled and the failure message
```
failure messages

ravigadde commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -139,14 +139,14 @@ type ExtenderArgs struct {
 	// Filtered set of nodes where the pod can be scheduled
 	Nodes apiv1.NodeList `json:"nodes,omitempty"`
-	// Filtered out nodes where the pod can't be scheduled and the extender that failed it
+	// Filtered out nodes where the pod can't be scheduled and the failure message
```
Same here

ravigadde commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -139,14 +139,14 @@ type ExtenderArgs struct {
 
-// FailedNodesMap represents the filtered out nodes map with node name and failed message
+// FailedNodesMap represents the filtered out nodes, with node name and failure message
 type FailedNodesMap map[string]string
```
node names and failure messages

ravigadde commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -139,14 +139,14 @@ type ExtenderArgs struct {
 }
 
-// FailedNodesMap represents the filtered out nodes map with node name and failed message
+// FailedNodesMap represents the filtered out nodes, with node name and failure message
```
Same here

ravigadde commented:
@ping035627 Couple of minor typos. Please address and rebase, I will ask @davidopp to add LGTM. Apologize for the delay in responding.

ping035627 commented:
@ravigadde Thanks for review, I've rebased and changed the comments, but the "Jenkins GKE smoke e2e" failed with the message "Build finished. No test results found.", can you help? thanks very much.

All components are up to date.
... in gke:kube-down()
... in gke:detect-project()
ERROR: (gcloud.container.clusters.delete) Some requests did not succeed:
 - args: ['ResponseError: code=404, message=The resource "projects/k8s-jkns-pr-gke/zones/us-central1-f/clusters/e2e-gke-agent-pr-1-0" was not found.\nNo cluster named \'e2e-gke-agent-pr-1-0\' in k8s-jkns-pr-gke.']
......
Master SSH not supported for gke
+ [[ 1 -eq 124 ]]
+ [[ 1 -eq 137 ]]
+ [[ 1 -ne 0 ]]
+ echo 'Build failed'
Build failed
+ echo 'Exiting with code: 1'
Exiting with code: 1

ravigadde commented:
@ping035627 not sure, could be a flake. Could you squash your commits? It should trigger another run.

ping035627 commented:
@ravigadde it's really a flake, the test run again and all passed, thanks very much. Now it needs lgtm.

ravigadde commented:
@davidopp Please add LGTM

davidopp commented:
@ravigadde you have verified that this will not cause backward compatibility problem (old/new version of scheduler with old/new version of extender) ?

ping035627 commented:
It's compatible because of json format, if one not  upgrade to new version, the new field  of  new version  will  be ignored,  and will  be processed filter as before.  We can verify it  with json.

ravigadde commented:
@davidopp Yes I have tested this new interface against the old extender and it works fine
